### PR TITLE
added optimized searching in partially sorted vector

### PR DESCRIPTION
**Pull request by sebastianbaginski**
@marcusnaslund review ?

I'm not sure if these functions should be called `first / lastIndex`, it is in fact `first not less` and `first greater`

### CODE REVIEW

File: source/system/structs/VectorList.ooc [hunk 1]
```diff
@@ -75,6 +75,55 @@ VectorList: class <T> extends List<T>{
...
+			if (!this isSorted(start, end, isLess))
+				raise("list must be partially sorted!")
+		count := end - start
+		current := start
+		while (count > 0) {
+			step := count / 2
+			current = start + step
+			compareResult: Bool
+			compareResult = isLess(this _vector[current]&, toCompare&)
+			if (compareResult) {
```
Comment (marcusnaslund) [inline source/system/structs/VectorList.ooc]: Can't you merge the above three lines? Rock shouldn't have any trouble with that, right? Right? RIGHT? *so sad*

**Comment by marcusnaslund** [comment]
Looks good and your names are fine.
Does the `current` variable really bring any benefit, though?

**Comment by sebastianbaginski** [comment]
I don't know, probably a leftover after `the thinking process` :) Will update shortly.

**Comment by sebastianbaginski** [comment]
Updated.

File: test/system/VectorListTest.ooc [hunk 2]
```diff
@@ -242,6 +243,34 @@ VectorListTest: class extends Fixture {
...
+		(intList, unorderedList, intComparator as Closure) free()
+
+		stringList := VectorList<String> new()
+		stringList add("a") . add("a") . add("b") . add("ab") . add("ba") . add("bb") .add("ccc") . add("ddd")
+		stringComparator := func (first, second: String*) -> Bool {
+			first@ size < second@ size
+		}
+		stringToCompare := "aa"
+		expect(stringList isSorted(0, stringList count, stringComparator), is true)
+		expect(stringList lowerBound(0, stringList count, stringToCompare, stringComparator), is equal to(3))
```
Comment (marcusnaslund) [inline test/system/VectorListTest.ooc]: Add a failing test (`equal to(-1)`) for the strings too.

**Comment by marcusnaslund** [comment]
After my additional comment, I'm happy.

File: test/system/VectorListTest.ooc (hunk 2 above)
Comment (sebastianbaginski) [inline test/system/VectorListTest.ooc]: Sure.

**Comment by sebastianbaginski** [comment]
Updated.

**Comment by marcusnaslund** [comment]
Perfection.

**Pull request merged**